ext/cl_khr_command_buffer.asciidoc

```diff
@@ -490,6 +490,19 @@ with the command-queue used on command-buffer creation. Where a _compatible_
 command-queue is defined as a command-queue with identical properties targeting
 the same device and in the same OpenCL context.
 
+While constructing a command-buffer it is valid for the user to interleave calls
+to the same queue which create commands, such as {clCommandNDRangeKernelKHR}, with
+queue submission calls, such as {clEnqueueNDRangeKernel} or
+{clEnqueueCommandBufferKHR}. That is, there is no effect on queue state from
+recording commands. The purpose of the queue parameter is to define the device
+and properties of the command, which are constant queries on the queue object.
+
+A command-buffer object should increment the reference count of OpenCL objects
+such as queues, buffers, images, and kernels referenced in commands recorded to
+the command-buffer. This enables correct behaviour of the command-buffer when the
+objects it uses have been released. On destruction of the command-buffer it should
+decrement these reference counts, allowing the referenced objects to be freed if
+appropriate.
```

Comment:
Actually one more comment with a few questions about this paragraph in general:

* Is this intended to increment application-visible reference count, or the internal reference count, or is either fine?  I know this is a complicated issue so I'm OK leaving this ambiguous if we need to.
* Should some of these "shoulds" turn into "musts"?
* I don't think the command-buffer should increment the reference count of any kernel arguments set on kernels recorded into the command-buffer (though it should retain or otherwise clone the kernel itself).  Do you agree?  If so, is it worth adding some non-normative text warning about this, similar to clSetKernelArg?

Reply:
Other parts of the spec talk about "attached" objects, e.g.

> After the device reference count becomes zero and all the objects attached to device (such as command-queues) are released, the device object is deleted.

I wonder if we should use the same term here, in which case the reference counting would fall out naturally?

> Implementations may also modify the reference count, e.g. to track attached objects...

The other term used in the spec is "associated with", but "attach" seems stronger and more appropriate for this use.

Reply:
ah, I didn't spot the consistent use of "attached". Rephrased this paragraph to use the term "attached" throughout

Reply:
Thanks!  I like the updated wording.